src/overlays/actors/ovl_Obj_Takaraya_Wall/z_obj_takaraya_wall.c

```diff
@@ -30,39 +50,450 @@ ActorInit Obj_Takaraya_Wall_InitVars = {
+
+// -10.0f height means the wall does not raise (the space is open)
+f32 sTakarayaWallWallHeights[TAKARAYA_WALL_ROWS][TAKARAYA_WALL_COLUMNS];
+
+// Note: returns true if the space is valid
+s32 ObjTakarayaWall_PosToSpace(Vec3f* pos, s32* row, s32* column) {
+    *row = ((s32)pos->x + (14 * 120)) / 120;
+    *column = (s32)pos->z / 120;
+
+    if ((*row < 0) || (*row >= TAKARAYA_WALL_ROWS) || (*column < 0) || (*column >= TAKARAYA_WALL_COLUMNS)) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * Checks if the given row and column is valid and is open, i.e does not raise up.
+ */
+s32 ObjTakarayaWall_SpaceOpen(s32 row, s32 column) {
+    if ((row < 0) || (row >= TAKARAYA_WALL_ROWS) || (column < 0) || (column >= TAKARAYA_WALL_COLUMNS) ||
+        (sTakarayaWallWallHeights[row][column] == 0.0f)) {
+        return false;
+    }
+
+    return true;
+}
+
+typedef enum TakarayaWallDirection {
+    /* 0 */ TAKARAYA_WALL_DIRECTION_BACK,
+    /* 1 */ TAKARAYA_WALL_DIRECTION_RIGHT,
+    /* 2 */ TAKARAYA_WALL_DIRECTION_FRONT,
+    /* 3 */ TAKARAYA_WALL_DIRECTION_LEFT
+} TakarayaWallDirection;
+
+/**
+ * Checks nearby spaces if the adjacent space or any of the three adjacent spaces to that
+ * space are closed (i.e the wall rises up).
+ */
+s32 ObjTakarayaWall_NearbySpacesClosed(s32 row, s32 column, TakarayaWallDirection direction) {
+    s32 row1;
+    s32 column1;
+    s32 adjacentSpaceRow;
+    s32 adjacentSpaceColumn;
+    s32 row2;
+    s32 column2;
+    s32 row3;
+    s32 column3;
+
+    if ((direction == TAKARAYA_WALL_DIRECTION_FRONT) || (direction == TAKARAYA_WALL_DIRECTION_BACK)) {
+        if (direction == TAKARAYA_WALL_DIRECTION_BACK) {
+            adjacentSpaceRow = row - 1;
+            row3 = row - 2;
+        } else { // TAKARAYA_WALL_DIRECTION_FRONT
+            adjacentSpaceRow = row + 1;
+            row3 = row + 2;
+        }
+        adjacentSpaceColumn = column;
+        column3 = column;
+        row1 = adjacentSpaceRow;
+        column1 = column + 1;
+        row2 = adjacentSpaceRow;
+        column2 = column - 1;
+    } else {
+        if (direction == TAKARAYA_WALL_DIRECTION_RIGHT) {
+            adjacentSpaceColumn = column - 1;
+            column3 = column - 2;
+        } else { // TAKARAYA_WALL_DIRECTION_LEFT
+            adjacentSpaceColumn = column + 1;
+            column3 = column + 2;
+        }
+        adjacentSpaceRow = row;
+        row3 = row;
+        row1 = row + 1;
+        column1 = adjacentSpaceColumn;
+        row2 = row - 1;
+        column2 = adjacentSpaceColumn;
+    }
+
+    if ((adjacentSpaceRow < 0) || (adjacentSpaceRow >= TAKARAYA_WALL_ROWS) || (adjacentSpaceColumn < 0) ||
+        (adjacentSpaceColumn >= TAKARAYA_WALL_COLUMNS) ||
+        (sTakarayaWallWallHeights[adjacentSpaceRow][adjacentSpaceColumn] == -10.0f) ||
+        ObjTakarayaWall_SpaceOpen(row1, column1) || ObjTakarayaWall_SpaceOpen(row2, column2) ||
+        ObjTakarayaWall_SpaceOpen(row3, column3)) {
+        return false;
+    }
```

Comment:
> As such, this function only returns true if "next cell" and "adjacent cells" are all closed.

Hence it it is checking if the nearby spaces are closed.
I'm not really a fan of `Can` in this situation. To go with the carving a path analogy though (which I really like by the way), perhaps something like `NearbyUncarved` would work better?

Reply:
Imo `NearbyUncarved` is definitely worse. The issue I have with it is that the maze isn't contained to just the 8x10 grid of editable cells, it extends "infinitely" in all directions. And remember, the default is closed/uncarved; if you test in a direction that leaves the grid, the checked cells will all be uncarved (because they can't be), but the function will return false because the next cell is off the grid. 

The other thing is that `NearbyUncarved` is less expressive of the function's purpose. It's primary goal is to determine if the cell to the up/right/down/left of (row, column) can potentially be turned into a path, the checking of nearby cells is implementation detail that doesn't need to be expressed in the function name. 

I also don't understand what's wrong with using `Can` here. Functions that return boolean states are typically written like a question, e.g. "Can I carve a path in this direction?" -> `CanCarvePathInDirection` and the function returns the answer to the question. `Can` also seems appropriate here because the function isn't changing state, so there aren't side effects to contend with

Reply:
The opposition had with `can` was it just didn't seem right when just considering within the maze proper itself (probably just my own issue as just having the adjacent cell valid and uncarved seems valid enough to be able to carve, why even consider cells next to that one?). However considering the invalid cells (which I admit hadn't crossed my mind), `can` does make more sense. I'd still say we can just shorten it to `ObjTakarayaWall_CanCarvePath`. The `InDirection` is already implied with the direction argument.